go/p2p/discovery/bootstrap/client.go

```diff
@@ -0,0 +1,162 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/discovery"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/p2p/rpc"
+)
+
+// ClientOptions are client's peer discovery options.
+type ClientOptions struct {
+	retentionPeriod time.Duration
+}
+
+// ClientOption is a client peer discovery option setter.
+type ClientOption func(opts *ClientOptions)
+
+// WithRetentionPeriod configures peer retention period.
+func WithRetentionPeriod(retentionPeriod time.Duration) ClientOption {
+	return func(opts *ClientOptions) {
+		opts.retentionPeriod = retentionPeriod
+	}
+}
+
+// client is an implementation of a peer discovery that fetches peers from the seed node and
+// advertises services.
+type client struct {
+	logger *logging.Logger
+
+	host host.Host
+	rc   rpc.Client
+
+	seed            peer.AddrInfo
+	retentionPeriod time.Duration
+
+	mu         sync.Mutex
+	peers      map[string][]peer.AddrInfo
+	timestamps map[string]time.Time
+}
+
+// NewClient creates a new bootstrap protocol client.
+func NewClient(h host.Host, seed peer.AddrInfo, opts ...ClientOption) discovery.Discovery {
+	l := logging.GetLogger("p2p/discovery/bootstrap").With(
+		"seed_id", seed.ID,
+		"seed_addrs", seed.Addrs,
+	)
+
+	do := ClientOptions{}
```

Comment:
Might make sense to define a default `ClientOptions` object, with some sane defaults (non-zero default `retentionPeriod`).